src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,375 @@
-    return jsg::JsValue(js.wrapReturningFunction(js.v8Context(),
-        [](jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
-          return js.error("WorkerRpc is unimplemented");
+  // Named intercept is enabled, this means we won't default to legacy behavior.
+  // The return value of the function is a promise that resolves once the remote returns the result
+  // of the RPC call.
+  return jsg::JsValue(js.wrapReturningFunction(js.v8Context(), [this, methodName=kj::str(name)]
+      (jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+        auto& ioContext = IoContext::current();
+        // Wait for the RPC to resolve and then process the result.
+        return js.wrapSimplePromise(ioContext.awaitIo(js, sendWorkerRpc(js, methodName, args),
+            [this] (jsg::Lock& js, WorkerRpc::ResponseAndResult result) -> jsg::Value {
+          return handleRpcResponse(js, kj::mv(result));
+        }));
+      }
+  ));
+}
+
+// The capability that lets us call remote methods over RPC.
+// The client capability is dropped after each callRequest().
+class JsRpcTargetImpl final : public rpc::JsRpcTarget::Server {
+public:
+  JsRpcTargetImpl(
+      kj::Own<kj::PromiseFulfiller<bool>> callFulfiller,
+      IoContext& ctx,
+      kj::Maybe<kj::StringPtr> entrypointName)
+      : callFulfiller(kj::mv(callFulfiller)), ctx(ctx), entrypointName(entrypointName) {}
+
+  // Handles the delivery of JS RPC method calls.
+  kj::Promise<void> call(CallContext callContext) override {
+    auto methodName = kj::heapString(callContext.getParams().getMethodName());
+    auto serializedArgs = callContext.getParams().getSerializedArgs().getV8Serialized().asBytes();
+
+    bool caughtException = false;
+    // Try to execute the requested method.
+    try {
+      caughtException = co_await ctx.run(
+          [this, methodName=kj::mv(methodName), serializedArgs = kj::mv(serializedArgs),
+           entrypointName = entrypointName, &callContext] (Worker::Lock& lock) ->
+              kj::Promise<bool> {
+
+        bool exceptionThrown = false;
+        auto reader = callContext.initResults(capnp::MessageSize { 4, 1 }).initResult();
+        jsg::Lock& js = lock;
+
+        // We will try to get the function, if we can't we'll throw an error to the client.
+        auto maybeFn = tryGetFn(lock, entrypointName, ctx, methodName);
+        KJ_IF_SOME(e, maybeFn.tryGet<kj::String>()) {
+          exceptionThrown = true;
+          reader.setV8Serialized(_::serializeV8(js, js.error(e)));
+          co_return exceptionThrown;
+        }
+
+        auto fn = KJ_REQUIRE_NONNULL(maybeFn.tryGet<v8::Local<v8::Function>>());
+        kj::Maybe<SerializedResult> eventualResult;
+
+        // We have a function, so let's call it, get the result, and serialize it for RPC.
+        // We account for both the JS method throwing and our serialization attempt failing.
+        if (fn->IsAsyncFunction()) {
```

Comment:
I suspect that using `IsAsyncFunction()` will not work in the case that the user gives us a function which is not declared `async` but happens to return a `Promise`. In this case, we still do want to wait for the `Promise` to resolve (we cannot serialize a `Promise` after all).

For example, imagine this code:

```
class C {
  foo() { return this.bar(DEFAULT_VALUE); }
  async bar(value) {
    ...
  }
}
```

Here, `foo()` is not marked `async`, but it returns a promise and it _ought_ to be treated as an async function.

So instead of using `IsAsyncFunction()`, we need to just call the function, and then check if the result is a `Promise`, and if so, wait on it.

Reply:
Merged them here: https://github.com/cloudflare/workerd/compare/810116c294c37febd3fdacb45d4391c1530266d4..eeb0822823fbd31e78653cd9005661cd0f29f9f1